### ISSUE TITLE
feat(trigger): odd trigger.

### ISSUE THREAD

**Issue by pmpetit** [pmpetit/pglinter#19]
# Description

fixed https://github.com/pmpetit/pglinter/pull/43

here is an example i can play with (it is the first time i see this kind of trigger)

in my mind a table and its triggers function should be the smallest as possible, and focused to its table, if possible. 
But maybe i am the only dba to be surprised by this trigger function. 

```sql
CREATE OR REPLACE FUNCTION odd.trigger_odd_revision_event()
    RETURNS trigger
    LANGUAGE plpgsql
AS $function$
DECLARE
    event_payload  JSONB;
    odd_ids varchar;
    id_odd int8;

BEGIN
    odd_ids := NULL;

    IF (UPPER(TG_TABLE_NAME) = 'T1') THEN
        select STRING_AGG(NEW.id::TEXT, ',' ) into odd_ids;
    END IF;
    IF (UPPER(TG_TABLE_NAME) = 'T2') THEN
        select STRING_AGG(distinct id::TEXT, ',' ) into odd_ids
        from (
                 select distinct b.id
                 from t1 b
                          inner join t4 lpm on lpm.id_surface = b.surface_id
                 where lpm.id_item_concept = NEW.id
                 union
                 select distinct b.id
                 from t1 b
                          inner join t2 pm on pm.parent_item_conception_id = b.item_id
                 where pm.id=NEW.id and b.surface_id is null
             ) as all_odds;
    END IF;
    IF (UPPER(TG_TABLE_NAME) = 'T3') THEN
        select STRING_AGG(distinct id::TEXT, ',' ) into odd_ids
        from (
                 select distinct b.id
                 from t1 b
                          inner join t4 lpm on lpm.id_surface = b.surface_id
                          inner join t2 pm on pm.id = lpm.id_item_concept
                          inner join t3 pa on (pa.id = NEW.id and pm.id = pa.parent_item_concept_id)
                 union
                 select distinct b.id
                 from t1 b
                          inner join t2 pm on pm.parent_item_conception_id = b.item_id
                          inner join t3 pa on (pa.id = NEW.id and pa.parent_item_concept_id = pm.id)
                 where b.surface_id is null
             ) as all_odds;
    END IF;
    IF (UPPER(TG_TABLE_NAME) = 'T4') THEN
        select STRING_AGG(distinct b.id::TEXT, ',' ) into odd_ids
        from t1 b
        where b.surface_id= COALESCE(NEW.id_surface,OLD.id_surface);
    END IF;

    IF (odd_ids is not null) THEN
        FOREACH id_odd IN array string_to_array(odd_ids::TEXT, ',')
            LOOP
                with concepts_agg_from_surface as (select b.id  as odd_id,
                                                          json_agg(json_build_object(
                                                                  'code', pm.item_code,
                                                                  'items', (select COALESCE(json_agg(json_build_object('code', pa.item_code)), '[]'::json) as items
                                                                            from t3 pa
                                                                            where pa.parent_item_concept_id = pm.id
                                                                            group by pa.parent_item_concept_id))) as concepts
                                                   from t1 b
                                                            inner join t4 lpm on lpm.id_surface = b.surface_id
                                                            inner join t2 pm   on (pm.id = lpm.id_item_concept and pm.parent_item_conception_id = b.item_id)
                                                   where b.id = id_odd
                                                   group by b.id),
                     concepts_agg_from_item as (select b.id as odd_id,
                                                        json_agg(json_build_object(
                                                                'code', pm.item_code,
                                                                'items', (select COALESCE(json_agg(json_build_object('code', pa.item_code)),'[]'::json) as items
                                                                          from t3 pa
                                                                          where pa.parent_item_concept_id = pm.id
                                                                          group by pa.parent_item_concept_id))) as concepts
                                                 from t1 b
                                                          inner join t2 pm on pm.parent_item_conception_id = b.item_id
                                                 where b.id = id_odd
                                                 group by b.id)
                -- objet final
                select jsonb_build_object(
                               'id', b.id,
                               'number', NULL,
                               'status', bs.code,
                               'concepts', (case
                                              when b.surface_id is not null then
                                                  (select COALESCE(m.concepts, '[]'::json) from concepts_agg_from_surface m where b.id = m.odd_id)
                                              else
                                                  (select COALESCE(m.concepts, '[]'::json) from concepts_agg_from_item m where b.id = m.odd_id)
                            end
                                   ),
                               'creation_date', b.creation_date,
                               'update_date', b.update_date,
                               'approval_date', NULL,
                               'comment', NULL,
                               'is_deleted', b.deleted_flag,
                               'creation_user', jsonb_build_object(
                                       'id', b.creation_user_id ,
                                       'name', creation_user.name
                                                ),
                               'update_user', jsonb_build_object(
                                       'id', b.update_user_id ,
                                       'name', update_user.name
                                              ),
                               'approval_user', null
                       )
                INTO event_payload
                from t1 b
                         inner join odd_status bs on bs.id = b.odd_status_id
                         left join fwkuser creation_user on b.creation_user_id = creation_user.userid
                         left join fwkuser update_user on b.update_user_id  = update_user.userid
                where b.id = id_odd;

                INSERT INTO relay_odd_revision_event
                ( creation_date, creator_id, event_status, sent_date, error_id, payload, odd_id, triggered_by)
                VALUES( now(), 'ODD', 'TO_SEND', NULL, NULL, event_payload,id_odd, TG_NAME);
            END LOOP;
    END IF;

    RETURN NULL;
END;
$function$
;
```  

if i were the coder, i should have splitted this function into several ones, to avoid the test on table_name. Each table could have its own trigger, instead of sharing it.

## Feature Definition

Maybe a rule could detect this kind of code, but i have no ideas about the way to do that.

**Comment by pmpetit** [comment]
I think a way to detect: if all tables that use a trigger use the same trigger function, then it could be a good starting point to look further.
